keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,510 @@
+removal of existing values in the
+target."(link)[https://tools.ietf.org/rfc/rfc7386.txt#:~:text=Null%20values%20in%20the%20merge]
+
+On the other hand, while `{field: null}` and `{}` probably have the same
+meaning, they do not "look" the same and are hard to compare for
+clients. They are also currently treated differently by Kubernetes for
+defaulting purposes.
+
+We propose the creation of a new CRD flag: `preserveNullValues` which
+will default to `true`. This flag will be used as a ratcheting mechanism
+as we migrate to this new semantic throughout all Kubernetes types
+(which would be how all types eventually work). When false, the
+apiserver will use new semantics and remove all null values from CR
+objects. When true, the existing and deprecated semantics will be kept.
+That new flag is following the same logic as "preserveUnknownFields" and
+the new operation will happen using the same mechanisms right after
+pruning.
+
+Note that we still consider `{}` (and 0, 0.0, "", false, []) to be
+different from `null`, so `{empty: {}}` is different from `{}` or from
+`{empty: null}` (but the latter two are considered equal).
+
+As described by @sttts on many occasions, the immutability is blocked
+because we don't know how to distinguish between null and
+unspecified. We believe that this mechanism will be able to unblock it.
+
+### Built-in Defaulting
+
+Because built-in types are serialized and deserialized as Go structures,
+there are semantics that are imposed on us. The two main ones are:
+1. We can't differentiate between null, unspecified and zero-value,
+2. non-pointer structs are always present (and hence have to be
+recursively defaulted), they also can't be omitEmpty'd.
+
+The first point is solved by the removed nulls described in the previous
+section (and invalid zero-values can't be present in objects), and the
+second point can be solved by forcing non-pointer structs to have a `{}`
+default. To address that, we're proposing that the `default` marker is
+forbidden on non-pointer structs (pointer structs can behave as
+expected) and the default field of the openapi is automatically set to
+`{}`. This only applies to OpenAPI generated from built-in types or from
+kubebuilder. CRDs can still have any defaulting they have for
+structures, but we strongly suggest that OpenAPI generated from Go types
+should follow that semantics.
+
+For example:
+```golang
+Type Root struct {
+  // Defaults on non-pointer structs are FORBIDDEN:
+  // +default={name: entry, number: 12}
+  Entry SubLevel
+}
+
+Type SubLevel struct {
+  Name string
+  Number int
+}
+```
+
+Computed defaults (through defaulting functions) will not be supported
+as a declarative marker and will continue being supported through the
+existing defaulting functions.
+
+### Risks and Mitigations
+
+The main risk is that we could end-up with different defaults for
+existing types which would be extremely confusing for our users.
```

Comment:
And the mitigation is what? Testing?

Reply:
+1 had the same question before. I doubt we do proper null, zero-value testing for defaults today. They just happen to work as they work due to unmarshalling.

Reply:
copying @jpbetz comment from the previous thread:

> When migrating from defaulting functions to +default tags, we could mitigate with proper testing. We should test that the defaults of existing types are working as expected anyway, so making sure that testing is in place seems reasonable and sufficient. Is there anything else about this risk we need to watch out for?

Reply:
Mentioned that double-checking tests and gaping the lack of coverage.